text/0042-secret-debug-display.md

```diff
@@ -0,0 +1,88 @@
++ Feature name: `secret-debug-display`
++ Start date: 2020-07-03
++ RFC PR: [iotaledger/bee-rfcs#42](https://github.com/iotaledger/bee-rfcs/pull/42)
++ Bee issue: [iotaledger/bee#100](https://github.com/iotaledger/bee/issues/100)
+
+# Summary
+
+This RFC introduces two [procedural macros](https://doc.rust-lang.org/reference/procedural-macros.html), `SecretDebug`
+and `SecretDisplay`, to derive the traits [Debug](https://doc.rust-lang.org/std/fmt/trait.Debug.html) and
+[Display](https://doc.rust-lang.org/std/fmt/trait.Display.html) for secret material types in order to avoid leaking
+their internals.
+
+# Motivation
+
+Secret materials, private keys for example, are types that are expected to remain private and be shared/leaked under no
+circumstances. However, these types may be wrapped in other higher level types, a wallet for example, that may want to
+implement `Debug` and/or `Display`. Not implementing these traits on the secret types would then not permit it.
+At the same time, secret materials should not be exposed by logging them. The consensus is then to implement `Debug` and
+`Display` but to not actually log the inner secret. To make things easier, two procedural macros are proposed to
+automatically derive  such implementations.
+
+# Detailed design
+
+## Crate
+
+Since procedural macros require a crate defined as `proc-macro = true`, a new dedicated crate has to be created.
+
+## Derivation
+
+This feature is expected to be used by deriving `SecretDebug` and/or `SecretDisplay`.
```

Comment:
I miss the actual trait definition for `SecretDebug` and `SecretDisplay` 🙂

Reply:
`SecretDebug` and `SecretDisplay` are not trait, they are procedural macros. And I am giving the definition of `SecretDebug`.

Reply:
But they produce the implementation of the traits `Debug` and `Display` for a type.

Reply:
I was just reading the documentation of `syn` where they define a `HeapSize` trait **and** a HeapSize proc macro. Here it doesn't seem to be necessary. I have never actually written proc macros myself yet.

Reply:
In their example they show how to add a bound (`HeapSize`) to a trait they are deriving the implementation. But we don't really need to add bounds to Display/Debug here.